lib/saluki-components/src/destinations/datadog_metrics/request_builder.rs

```diff
@@ -243,11 +256,79 @@ where
 
         self.compression_estimator.reset();
 
+        // Clear scratch buf data.
+        let scratch_buf = std::mem::take(&mut self.scratch_buf);
+        let scratch_buf_lens = std::mem::take(&mut self.scratch_buf_lens);
+
         let new_compressor = create_compressor(&self.buffer_pool).await;
         let mut compressor = std::mem::replace(&mut self.compressor, new_compressor);
-        compressor.shutdown().await.context(Io)?;
+        if let Err(e) = compressor.shutdown().await.context(Io) {
+            return vec![Err(e)];
+        }
+
         let buffer = compressor.into_inner();
 
+        let compressed_len = buffer.len();
+        let compressed_limit = self.endpoint.compressed_size_limit();
+        if compressed_len > compressed_limit {
+            return self.split_request_efficient(scratch_buf, scratch_buf_lens).await;
+        }
+
+        debug!(endpoint = ?self.endpoint, uncompressed_len, compressed_len, "Flushing request.");
+
+        vec![self.create_request(buffer).map(|req| (metrics_written, req))]
+    }
+
+    async fn split_request_efficient(
+        &self, scratch_buf: Vec<u8>, scratch_buf_lens: Vec<usize>,
+    ) -> Vec<Result<(usize, Request<ReplayBody<ChunkedBuffer<O>>>), RequestBuilderError>> {
+        let mut requests = Vec::new();
+
+        if scratch_buf_lens.is_empty() {
+            return requests;
+        }
+
+        let pivot = scratch_buf_lens.len() / 2;
+        let first_half: usize = scratch_buf_lens.iter().take(pivot + 1).sum();
+
+        let mut compressor_half_one = create_compressor(&self.buffer_pool).await;
+
+        if first_half <= scratch_buf.len() {
```

Comment:
I'm not actually sure you need these conditionals (this one and the similar one further down) unless you're incorrectly generating the length of each write way further up in `encode`.

You should do this once, perhaps with an `assert!`, right below my proposed `scratch_buf_pivot`, and just assert that `scratch_buf_pivot < scratch_buf.len()`, since that's all we need to do to ensure we never generate an invalid slice reference.